src/EventListener/AbstractUpdateListener.php

```diff
@@ -5,78 +5,45 @@
-use Doctrine\Common\Persistence\ObjectManager;
 use Doctrine\ORM\EntityManagerInterface;
 use Setono\SyliusRedirectPlugin\Decider\AutomaticRedirectCreationDeciderInterface;
-use Setono\SyliusRedirectPlugin\Factory\RedirectFactoryInterface;
-use Setono\SyliusRedirectPlugin\Finder\RemovableRedirectFinderInterface;
-use Setono\SyliusRedirectPlugin\Model\RedirectInterface;
+use Setono\SyliusRedirectPlugin\Exception\SlugUpdateHandlerValidationException;
+use Setono\SyliusRedirectPlugin\SlugUpdateHandler\SlugUpdateHandlerCommand;
+use Setono\SyliusRedirectPlugin\SlugUpdateHandler\SlugUpdateHandlerInterface;
 use Sylius\Bundle\ResourceBundle\Event\ResourceControllerEvent;
 use Sylius\Component\Resource\Model\SlugAwareInterface;
 use Symfony\Component\Validator\ConstraintViolationInterface;
-use Symfony\Component\Validator\Validator\ValidatorInterface;
 use Webmozart\Assert\Assert;
 
 abstract class AbstractUpdateListener
 {
-    /** @var ValidatorInterface */
-    protected $validator;
-
     /** @var ManagerRegistry */
     protected $managerRegistry;
 
-    /** @var RemovableRedirectFinderInterface */
-    protected $removableRedirectFinder;
-
-    /** @var RedirectFactoryInterface */
-    protected $redirectFactory;
-
     /** @var AutomaticRedirectCreationDeciderInterface */
     protected $automaticRedirectCreationDecider;
 
-    /** @var array */
-    protected $validationGroups;
+    /** @var SlugUpdateHandlerInterface */
+    private $slugUpdateHandler;
 
-    /** @var string */
-    protected $class;
-
-    /** @var ObjectManager|null */
-    private $manager;
+    /** @var EntityManagerInterface[] */
+    private $managers = [];
 
     public function __construct(
-        ValidatorInterface $validator,
         ManagerRegistry $managerRegistry,
-        RemovableRedirectFinderInterface $removableRedirectFinder,
-        RedirectFactoryInterface $redirectFactory,
         AutomaticRedirectCreationDeciderInterface $automaticRedirectCreationDecider,
-        array $validationGroups,
-        string $class
+        SlugUpdateHandlerInterface $slugUpdateHandler
     ) {
-        $this->validator = $validator;
         $this->managerRegistry = $managerRegistry;
-        $this->removableRedirectFinder = $removableRedirectFinder;
-        $this->redirectFactory = $redirectFactory;
         $this->automaticRedirectCreationDecider = $automaticRedirectCreationDecider;
-        $this->validationGroups = $validationGroups;
-        $this->class = $class;
+        $this->slugUpdateHandler = $slugUpdateHandler;
     }
 
-    abstract protected function getPostName(): string;
-
-    abstract protected function createRedirect(
-        SlugAwareInterface $slugAware,
-        string $source,
-        string $destination,
-        bool $permanent = true,
-        bool $only404 = true
-    ): RedirectInterface;
-
-    protected function getPrevious(SlugAwareInterface $slugAware): array
+    private function getPrevious(SlugAwareInterface $slugAware): array
```

Comment:
I would make this protected to allow better overriding. It might be useful to be able to override this behaviour. ie. fetching it from another data source than doctrine.